### PR TITLE
Flesh out redpallas, direct port of redjubjub

### LINKED ISSUE #2044: Replace redpallas stub in zebra-chain with full implementation
**Is your feature request related to a problem? Please describe.**
To suffice types in the sapling->orchard port in zebra-chain, we created [`redpallas`](https://github.com/ZcashFoundation/zebra/blob/b8fab0464695bc7ffd63f947d2a72345447c9f99/zebra-chain/src/primitives/redpallas.rs) types but they are functionally incomplete. They need to be functional to verify SpendAuth and Binding signatures for NU5. 

**Describe the solution you'd like**
@str4d has  implemented [`reddsa` as a fork of `redjubjub`](https://github.com/str4d/redjubjub/tree/orchard) to abstract the common types and traits into a dependency for the [`redpallas` types in the `orchard` crate](https://github.com/zcash/orchard/blob/4b05c20a2ddac0eee9fbf9153cb081442124aec4/src/primitives/redpallas.rs).  The reddsa orchard branch there defines the Orchard basepoints, which is not my favorite.

Should the `redjubjub` crate turn into something a more shared `reddsa`, that includes the `redjubjub` and `reddsa`? With FROST, it may be annoying to have 3 crates to juggle: `reddsa` as a dependency, `redjubjub`, `redpallas`, with boilerplate across them. 

Or should we just duplicate where necessary?

### PR DESCRIPTION
**Pull request by dconnolly**
## Motivation

Finishing Transaction::V5 requires a `redpallas::Signature` type, which was not part of the initial stub-out of the `redpallas` types in the sapling->orchard port in #1885.

## Solution

Finish the direct `redjubjub` -> `redpallas` port for single signature verification, modulo the batching and the multiscalar multiplication required for that, as the math needs updating or replacing completely.

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Unit Tests and Property Tests

## Review

This unblocks @oxarbitrage 

## Related Issues

This technically fulfills #2044, but I don't know if we want it to be the long-term solution. @mpguerra  

## Follow Up Work

The batch math: https://github.com/ZcashFoundation/zebra/issues/2098

### CODE REVIEW

**Comment by teor2345** [comment]
> This technically fulfills #2044, but I don't know if we want it to be the long-term solution. @mpguerra

I think it's fine to close that ticket for now, and open an optional cleanup ticket for later.

File: zebra-chain/src/primitives/redpallas/signing_key.rs [hunk 1]
```diff
@@ -30,10 +38,88 @@ impl<T: SigType> TryFrom<[u8; 32]> for SigningKey<T> {
...
+            .update(&random_bytes[..])
+            .update(&self.pk.bytes.bytes[..]) // XXX ugly
+            .update(msg)
+            .finalize();
+
+        let r_bytes = pallas::Affine::from(T::basepoint() * nonce).to_bytes();
+
+        let c = HStar::default()
+            .update(&r_bytes[..])
+            .update(&self.pk.bytes.bytes[..]) // XXX ugly
```
Comment (teor2345) [inline zebra-chain/src/primitives/redpallas/signing_key.rs:113]: Try:
```suggestion
            .update(self.pk.bytes.bytes.as_slice())
            .update(msg)
            .finalize();

        let r_bytes = pallas::Affine::from(T::basepoint() * nonce).to_bytes();

        let c = HStar::default()
            .update(&r_bytes[..])
            .update(self.pk.bytes.bytes.as_slice())
```

File: zebra-chain/src/primitives/redpallas/constants.rs [hunk 2]
```diff
@@ -11,12 +11,12 @@
...
+    24, 141, 59, 6, 214, 241, 40, 179, 35, 85, 183,
 ];
 
 /// The byte-encoding of the basepoint for `BindingSig` on the Pallas curve.
 pub const BINDINGSIG_BASEPOINT_BYTES: [u8; 32] = [
-    48, 181, 242, 170, 173, 50, 86, 48, 188, 221, 219, 206, 77, 103, 101, 109, 5, 253, 28, 194,
-    208, 55, 187, 83, 117, 182, 233, 109, 158, 1, 161, 215,
+    145, 90, 60, 136, 104, 198, 195, 14, 47, 128, 144, 238, 69, 215, 110, 64, 72, 32, 141, 234, 91,
+    35, 102, 79, 187, 9, 164, 15, 85, 68, 244, 7,
 ];
```
Comment (teor2345) [inline zebra-chain/src/primitives/redpallas/constants.rs]: These seem to duplicate the `lazy_static`s in:
https://github.com/ZcashFoundation/zebra/blob/566a10151ca8ab40b4568ab75c2bfaa5881618b6/zebra-chain/src/orchard/commitment.rs#L324

File: zebra-chain/src/primitives/redpallas/tests/prop.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+struct SignatureCase<T: SigType> {
+    msg: Vec<u8>,
+    sig: Signature<T>,
+    pk_bytes: VerificationKeyBytes<T>,
+    is_valid: bool,
+}
+
+/// A modification to a test-case.
+#[derive(Copy, Clone, Debug)]
+enum Tweak {
```
Comment (teor2345) [inline zebra-chain/src/primitives/redpallas/tests/prop.rs:22]: 👀

File: zebra-chain/src/primitives/redpallas/tests/prop.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+proptest! {
+    #[test]
+    fn tweak_signature(
+        tweaks in prop::collection::vec(tweak_strategy(), (0,5)),
+        rng_seed in any::<u64>(),
+    ) {
+        // Use a deterministic RNG so that test failures can be reproduced.
+        // Seeding with 64 bits of entropy is INSECURE and this code should
+        // not be copied outside of this test!
+        let mut rng = ChaChaRng::seed_from_u64(rng_seed);
```
Comment (teor2345) [inline zebra-chain/src/primitives/redpallas/tests/prop.rs]: Sometimes it's safer just to use the correct code :-)

File: zebra-chain/src/primitives/redpallas/verification_key.rs [hunk 5]
```diff
@@ -115,3 +99,81 @@ impl<T: SigType> TryFrom<[u8; 32]> for VerificationKey<T> {
...
+        };
+        VerificationKey { point, bytes }
+    }
+
+    /// Verify a purported `signature` over `msg` made by this verification key.
+    // This is similar to impl signature::Verifier but without boxed errors
+    pub fn verify(&self, msg: &[u8], signature: &Signature<T>) -> Result<(), Error> {
+        let c = HStar::default()
+            .update(&signature.r_bytes[..])
+            .update(&self.bytes.bytes[..]) // XXX ugly
```
Comment (teor2345) [inline zebra-chain/src/primitives/redpallas/verification_key.rs:133]: Try:
```suggestion
            .update(self.bytes.bytes.as_slice())
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks great.

I had a few suggestions, but they're all minor.

File: zebra-chain/src/primitives/redpallas/signature.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+//
+// Authors:
+// - Henry de Valence <hdevalence@hdevalence.ca>
+// - Deirdre Connolly <deirdre@zfnd.org>
+
+use std::marker::PhantomData;
+
+use super::SigType;
+
+/// A RedJubJub signature.
```
Comment (conradoplg) [inline zebra-chain/src/primitives/redpallas/signature.rs]:
```suggestion
/// A RedPallas signature.
```

**Review by conradoplg — CHANGES_REQUESTED** [review]
On a first pass I just found a small issue in a comment.

Should I make a more in-depth review of the crypto part? (It should take a couple days to get familiar with RedDSA)

File: zebra-chain/src/primitives/redpallas/signing_key.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-chain/src/primitives/redpallas/signing_key.rs:113]: Ah I would but it's not available on stable, even with the feature flag

File: zebra-chain/src/primitives/redpallas/constants.rs (hunk 2 above)
Comment (dconnolly) [reply zebra-chain/src/primitives/redpallas/constants.rs]: This is interesting; so what I may do is swap what we have, but keep the test cases that assert that the basepoints calculated for spendauth and binding and `lazy_static!`'d match the bytes we expect.

Comment (dconnolly) [reply zebra-chain/src/primitives/redpallas/constants.rs]: > These seem to duplicate the `lazy_static`s in:
> https://github.com/ZcashFoundation/zebra/blob/566a10151ca8ab40b4568ab75c2bfaa5881618b6/zebra-chain/src/orchard/commitment.rs#L324

Oh yes the Binding Sig basepoint is a duplicate, correct.  The SpendAuth basepoint is distinct from V in the value commitments

Comment (dconnolly) [reply zebra-chain/src/primitives/redpallas/constants.rs]: So originally, had in mind that any `redpallas` would become a separate crate/repo like `redjubjub` is, so deduping some of these constants may not help there, it wouldn't be hard to split back out again, but

File: zebra-chain/src/primitives/redpallas/tests/prop.rs (hunk 4 above)
Comment (dconnolly) [reply zebra-chain/src/primitives/redpallas/tests/prop.rs]: Fixed! 7b3ad86a

File: zebra-chain/src/primitives/redpallas/constants.rs (hunk 2 above)
Comment (teor2345) [reply zebra-chain/src/primitives/redpallas/constants.rs]: Whatever you think is best here, as long as there are tests to make sure they all match

File: zebra-chain/src/primitives/redpallas/signing_key.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/primitives/redpallas/signing_key.rs:113]: Oh right that's the one as_slice impl that isn't stable.

File: zebra-chain/src/primitives/redpallas/tests/prop.rs (hunk 4 above)
Comment (teor2345) [reply zebra-chain/src/primitives/redpallas/tests/prop.rs]: Yep you know how people love to copy code (and not read or copy comments) 🙃

**Review by teor2345 — DISMISSED** [review]
This looks good, happy to merge if you're happy with it

File: zebra-chain/src/primitives/redpallas/constants.rs (hunk 2 above)
Comment (dconnolly) [reply zebra-chain/src/primitives/redpallas/constants.rs]: If there isn't much movement to extract redpallas in the near future, I might reorganize several of these shared constants or make them more pub(crate) so we can check cross-compatibility.

**Pull request merged**